src/actix/api/snapshot_api.rs

```diff
@@ -333,24 +333,34 @@ async fn upload_shard_snapshot(
     query: web::Query<SnapshotUploadingParam>,
     MultipartForm(form): MultipartForm<SnapshottingForm>,
 ) -> impl Responder {
+    let (collection, shard) = path.into_inner();
     let SnapshotUploadingParam { wait, priority } = query.into_inner();
 
-    let future = async move {
-        let (collection, shard) = path.into_inner();
-        let collection = toc.get_collection(&collection).await?;
-        collection.assert_shard_exists(shard).await?;
+    let future = cancel::future::spawn_cancel_on_drop(move |cancel| async move {
+        let future = async {
+            let collection = toc.get_collection(&collection).await?;
+            collection.assert_shard_exists(shard).await?;
+
+            Result::<_, helpers::HttpError>::Ok(collection)
+        };
+
+        let collection = cancel::future::cancel_on_token(cancel.clone(), future).await??;
```

Comment:
Good point, but as far as I understand, `upload_shard_snapshot` is only called after file is fully uploaded (because `form` contains a ready-to-use `TempFile`). So if you cancel the request during file upload, request should be cancelled and partially uploaded file would be removed.

Reply:
Brilliant!